### DISCUSSION TITLE
sftpgo.json and users table encryption

### DISCUSSION THREAD

**Discussion by saul-hu** [drakkan/sftpgo#446] [Ideas]
Hi ,
    In order to prevent password leakage and data security, I think the following settings need to be encrypted. What do you think?
1." data_ provider" > "Password" in sftpgo.json file
2. The SAS URL of the users table

**Comment by drakkan** [comment]
Hi,

the SAS URL could be encrypted as we already do the the account key (and so it can be stored using an external KMS system). Regarding the configurations within `sftpgo.json` I don't think we can encrypt them: how will the decryption key be read? Anyone accessing the decryption key could decrypt the password as well. We should also provide a CLI tool or similar to encrypt the password.

We could add support for loading the configuration file from Vault or other supported KMS but you need to provide the KMS URL and credentials in some way (for example via env vars). Do you use a KMS?

However protecting the configuration files at filesystem levels seems an accepted standard, please take a look here:

https://docs.aws.amazon.com/it_it/general/latest/gr/aws-access-keys-best-practices.html

if you are on docker you can also user [docker secrets](https://docs.docker.com/engine/swarm/secrets/)

You can also leave the dataprovider password empty in your configuration file and set it using an environment variable

**Reply by saul-hu** [reply]
Hi，
Thank you very much for your reply and suggestions.
We didn't use kms.
Thank you for your advice on passwd in sftpgo. JSON.
Can SAS URL encryption be implemented in the next version?

**Reply by drakkan** [reply]
Hi,

a [sponsorship](https://github.com/sponsors/drakkan) could help. 

I will try to add this feature anyway, I really enjoy working on SFTPGo but it is getting really challenging. For long term project sustainability I need sponsorships as I don't use SFTPGo in production, this is totally a hobby project, and cannot continue to refuse other paid jobs forever (or ship them with delays) to have enough time to work on SFTPGo. Thanks for understanding

**Reply by drakkan** [reply]
Please give a try to https://github.com/drakkan/sftpgo/commit/9d3d7db29ce325a34839e45c2722a026cb28401c
